File: golang/438.find-all-anagrams-in-a-string.go
func findAnagrams(s string, p string) []int {
	n := len(s)
	m := len(p)

	indices := make([]int, 0)

	pattern := [26]int{}
	cur := [26]int{}

	for i := range m {
		pattern[p[i]-'a']++
	}

	for i := range n {
		cur[s[i]-'a']++

		if i-m >= 0 {
			cur[s[i-m]-'a']--
		}

		if i+1-m >= 0 {
			if pattern == cur {
				indices = append(indices, i+1-m)
			}
		}

	}

	return indices
}

File: python/438.find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        """using hash generator function for window.
        """
        def gen(st: str, window: Optional[int]=None) -> Generator[int, None, None]:
            if window is None:
                window = len(st)
            n = len(st)
            h = sum(map(hash, st[:window]))
            yield h
            for i in range(window, n):
                h += hash(st[i]) - hash(st[i - window])
                yield h

        res = []

        len_s, len_p = len(s), len(p)
        if len_p > len_s:
            return res
        
        pattern = next(gen(p))
        return [i for i, curr in enumerate(gen(s, len_p)) if curr == pattern]
